The Chinook record store has just signed a deal with a new record label, and you've been tasked with selecting the first three albums that will be added to the store, from a list of four. All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:

|Artist Name|Genre|
|-----|-----|
|Regal|Hip-Hop|
|Red Tone|Punk|
|Meteor and the Girls|Pop|
|Slim Jim Bites|Blues|

The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, so we're interested in finding out which genres sell the best in the USA.

In order to do so, we will query the chinook database to find out which genres are the most popular in the USA, and choose the corresponding artist to add their album to the store.

In [77]:
%%html
<style>
.nbviewer div.output_area {
  overflow-y: auto;
  max-height: 200px; /* or value of your choosing */
}
</style>

In [2]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

In [16]:
%%sql

SELECT *
FROM sqlite_master
WHERE type in ('table', 'view')

 * sqlite:///chinook.db
Done.


[('table', 'album', 'album', 2, 'CREATE TABLE [album]\n(\n    [album_id] INTEGER PRIMARY KEY NOT NULL,\n    [title] NVARCHAR(160)  NOT NULL,\n    [artist_id] INTEGER  NOT NULL,\n    FOREIGN KEY ([artist_id]) REFERENCES [artist] ([artist_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'artist', 'artist', 3, 'CREATE TABLE [artist]\n(\n    [artist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'customer', 'customer', 4, 'CREATE TABLE [customer]\n(\n    [customer_id] INTEGER PRIMARY KEY NOT NULL,\n    [first_name] NVARCHAR(40)  NOT NULL,\n    [last_name] NVARCHAR(20)   ... (276 characters truncated) ...  [support_rep_id] INTEGER,\n    FOREIGN KEY ([support_rep_id]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'employee', 'employee', 5, 'CREATE TABLE [employee]\n(\n    [employee_id] INTEGER PRIMARY KEY NOT NULL,\n    [last_name] NVARCHAR(20)  NOT NULL,\n    [first_name] NVARCHAR(20)   ... (310 characters truncated) ... 4),\n    [email] NVARCHAR(60),\n    FOREIGN KEY ([reports_to]) REFERENCES [employee] ([employee_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'genre', 'genre', 6, 'CREATE TABLE [genre]\n(\n    [genre_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'invoice', 'invoice', 7, 'CREATE TABLE [invoice]\n(\n    [invoice_id] INTEGER PRIMARY KEY NOT NULL,\n    [customer_id] INTEGER  NOT NULL,\n    [invoice_date] DATETIME  NOT NUL ... (195 characters truncated) ... tal] NUMERIC(10,2)  NOT NULL,\n    FOREIGN KEY ([customer_id]) REFERENCES [customer] ([customer_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'invoice_line', 'invoice_line', 8, 'CREATE TABLE [invoice_line]\n(\n    [invoice_line_id] INTEGER PRIMARY KEY NOT NULL,\n    [invoice_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT ... (161 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'media_type', 'media_type', 9, 'CREATE TABLE [media_type]\n(\n    [media_type_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist', 'playlist', 10, 'CREATE TABLE [playlist]\n(\n    [playlist_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(120)\n)'),
 ('table', 'playlist_track', 'playlist_track', 11, 'CREATE TABLE [playlist_track]\n(\n    [playlist_id] INTEGER  NOT NULL,\n    [track_id] INTEGER  NOT NULL,\n    CONSTRAINT [pk_playlist_track] PRIMARY ... (114 characters truncated) ...  DELETE NO ACTION ON UPDATE NO ACTION,\n    FOREIGN KEY ([track_id]) REFERENCES [track] ([track_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('table', 'track', 'track', 13, 'CREATE TABLE [track]\n(\n    [track_id] INTEGER PRIMARY KEY NOT NULL,\n    [name] NVARCHAR(200)  NOT NULL,\n    [album_id] INTEGER,\n    [media_type_ ... (377 characters truncated) ... ON ON UPDATE NO ACTION,\n    FOREIGN KEY ([media_type_id]) REFERENCES [media_type] ([media_type_id]) \n\t\tON DELETE NO ACTION ON UPDATE NO ACTION\n)'),
 ('view', 'tracks_sold_usa', 'tracks_sold_usa', 0, "CREATE VIEW tracks_sold_usa AS \n    SELECT *\n    FROM invoice i\n    INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id\n    WHERE billing_country = 'USA'")]

In [43]:
%%sql

/*Find number of tracks sold in the USA for each genre*/

/*Create view of sales that had USA as billing country. 
The information includes customer_id, invoice_id, track_id, quantity, track_name, genre_id, and genre_name*/

DROP VIEW IF EXISTS tracks_sold_usa;

CREATE VIEW tracks_sold_usa AS 
    SELECT
        i.customer_id customer_id,
        i.invoice_id,
        il.track_id,
        t.name track_name,
        il.quantity,
        t.genre_id,
        g.name genre_name      
    FROM invoice i
    INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id
    INNER JOIN track t ON il.track_id = t.track_id
    INNER JOIN genre g ON g.genre_id = t.genre_id
    WHERE billing_country = 'USA';

WITH total_sold AS (
    SELECT SUM(quantity)
    FROM tracks_sold_usa)   


SELECT COUNT(quantity) number_sold, 
       100*COUNT(quantity)/CAST((SELECT SUM(quantity)
                        FROM tracks_sold_usa) AS FLOAT) percentage_sold,
       genre_name
FROM tracks_sold_usa
GROUP BY genre_name
ORDER BY number_sold DESC;

 * sqlite:///chinook.db
Done.
Done.
Done.


[(561, 53.37773549000951, 'Rock'),
 (130, 12.369172216936251, 'Alternative & Punk'),
 (124, 11.798287345385347, 'Metal'),
 (53, 5.042816365366318, 'R&B/Soul'),
 (36, 3.4253092293054235, 'Blues'),
 (35, 3.330161750713606, 'Alternative'),
 (22, 2.093244529019981, 'Pop'),
 (22, 2.093244529019981, 'Latin'),
 (20, 1.9029495718363463, 'Hip Hop/Rap'),
 (14, 1.3320647002854424, 'Jazz'),
 (13, 1.236917221693625, 'Easy Listening'),
 (6, 0.570884871550904, 'Reggae'),
 (5, 0.47573739295908657, 'Electronica/Dance'),
 (4, 0.38058991436726924, 'Classical'),
 (3, 0.285442435775452, 'Heavy Metal'),
 (2, 0.19029495718363462, 'Soundtrack'),
 (1, 0.09514747859181731, 'TV Shows')]

From the table above, we see that the ranking of the 4 genres in the USA in terms of the number of tracks sold is as follows:

|Genre|Rank|Number Sold|Percentage of USA's total sales|
|-----|-----|-----|-----|
|Alternative & Punk|2|130|12.4|
|Blues|5|36|3.43|
|Pop|7|22|2.09|
|Hip Hop/Rap|9|20|1.90|

Based on this alone, we should choose Red Tone, Meteor and the Girls, and Slim Jim Bites to add their albums to the store, as their albums would help us to appeal to the widest possible crowd that is interested in the Alt & Punk, Pop, and Blues scenes respectively.


Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. In order to see if there any differences in the performances of the various sales support agents, we will analyse the purchases of the customers they are in charge of.

To begin, we first find out which of the employees are Sales Support Agents and create a View of them.

In [50]:
%%sql

DROP VIEW IF EXISTS sales_support;

CREATE VIEW sales_support AS
    SELECT *
    FROM employee
    WHERE title = "Sales Support Agent";
    
SELECT * FROM sales_support

 * sqlite:///chinook.db
Done.
Done.
Done.


[(3, 'Peacock', 'Jane', 'Sales Support Agent', 2, '1973-08-29 00:00:00', '2017-04-01 00:00:00', '1111 6 Ave SW', 'Calgary', 'AB', 'Canada', 'T2P 5M5', '+1 (403) 262-3443', '+1 (403) 262-6712', 'jane@chinookcorp.com'),
 (4, 'Park', 'Margaret', 'Sales Support Agent', 2, '1947-09-19 00:00:00', '2017-05-03 00:00:00', '683 10 Street SW', 'Calgary', 'AB', 'Canada', 'T2P 5G3', '+1 (403) 263-4423', '+1 (403) 263-4289', 'margaret@chinookcorp.com'),
 (5, 'Johnson', 'Steve', 'Sales Support Agent', 2, '1965-03-03 00:00:00', '2017-10-17 00:00:00', '7727B 41 Ave', 'Calgary', 'AB', 'Canada', 'T3B 1Y7', '1 (780) 836-9987', '1 (780) 836-9543', 'steve@chinookcorp.com')]

Next, we calculate the total dollar each customer has spent on the Chinook store. We do this by checking the total amount they have spent across all their invoices.

In [79]:
%%sql

SELECT c.customer_id, (c.first_name || ' ' || c.last_name) customer_name, SUM(i.total) total_spent, c.support_rep_id, (e.first_name || ' ' || e.last_name) employee_name, e.hire_date
FROM customer c
INNER JOIN invoice i ON c.customer_id = i.customer_id
INNER JOIN employee e ON c.support_rep_id = e.employee_id
GROUP BY i.customer_id
ORDER BY i.customer_id;

 * sqlite:///chinook.db
Done.


[(1, 'Luís Gonçalves', 108.89999999999998, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (2, 'Leonie Köhler', 82.17, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (3, 'François Tremblay', 99.99, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (4, 'Bjørn Hansen', 72.27000000000001, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (5, 'František Wichterlová', 144.54000000000002, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (6, 'Helena Holý', 128.7, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (7, 'Astrid Gruber', 69.3, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (8, 'Daan Peeters', 60.38999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (9, 'Kara Nielsen', 37.61999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (10, 'Eduardo Martins', 60.39, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (11, 'Alexandre Rocha', 69.3, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (12, 'Roberto Almeida', 82.17, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (13, 'Fernanda Ramos', 106.91999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (14, 'Mark Philips', 29.699999999999996, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (15, 'Jennifer Peterson', 66.33, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (16, 'Frank Harris', 74.25, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (17, 'Jack Smith', 98.01, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (18, 'Michelle Brooks', 79.2, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (19, 'Tim Goyer', 54.449999999999996, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (20, 'Dan Miller', 95.03999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (21, 'Kathy Chase', 91.08000000000001, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (22, 'Heather Leacock', 92.07000000000001, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (23, 'John Gordon', 66.33, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (24, 'Frank Ralston', 71.28, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (25, 'Victor Stevens', 76.23, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (26, 'Richard Cunningham', 86.13000000000002, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (27, 'Patrick Gray', 84.14999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (28, 'Julia Barnett', 72.27, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (29, 'Robert Brown', 40.59, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (30, 'Edward Francis', 91.08, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (31, 'Martha Silk', 62.370000000000005, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (32, 'Aaron Mitchell', 70.28999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (33, 'Ellie Sullivan', 75.24000000000001, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (34, 'João Fernandes', 102.96000000000001, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (35, 'Madalena Sampaio', 82.17, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (36, 'Hannah Schneider', 85.14, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (37, 'Fynn Zimmermann', 94.05000000000001, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (38, 'Niklas Schröder', 73.25999999999999, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (39, 'Camille Bernard', 79.2, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (40, 'Dominique Lefebvre', 72.27, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (41, 'Marc Dubois', 64.35, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (42, 'Wyatt Girard', 99.99, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (43, 'Isabelle Mercier', 73.25999999999999, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (44, 'Terhi Hämäläinen', 79.2, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (45, 'Ladislav Kovács', 78.21, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (46, "Hugh O'Reilly", 114.83999999999997, 3, 'Jane Peacock', '1973-08-29 00:00:00'),
 (47, 'Lucas Mancini', 50.49, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (48, 'Johannes Van der Berg', 65.34, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (49, 'Stanisław Wójcik', 76.22999999999999, 4, 'Margaret Park', '1947-09-19 00:00:00'),
 (50, 'Enrique Muñoz', 98.01, 5, 'Steve Johnson', '1965-03-03 00:00:00'),
 (51, 'Joakim

Next, we check the total amount spent by all the customers belonging to each SSA.

In [89]:
%%sql

SELECT support_rep_id ssa_id, employee_name ssa_name, hire_date ssa_hiredate, SUM(total_spent) total_of_ssa
FROM
    (SELECT c.customer_id, (c.first_name || ' ' || c.last_name) customer_name, SUM(i.total) total_spent, c.support_rep_id, (e.first_name || ' ' || e.last_name) employee_name, e.hire_date
    FROM customer c
    INNER JOIN invoice i ON c.customer_id = i.customer_id
    INNER JOIN employee e ON c.support_rep_id = e.employee_id
    GROUP BY i.customer_id
    ORDER BY i.customer_id)
GROUP BY ssa_id



 * sqlite:///chinook.db
Done.


[(3, 'Jane Peacock', '2017-04-01 00:00:00', 1731.5099999999998),
 (4, 'Margaret Park', '2017-05-03 00:00:00', 1584.0000000000002),
 (5, 'Steve Johnson', '2017-10-17 00:00:00', 1393.92)]

From the above table, we see that Jane Peacock had the greatest sales numbers, followed by Margaret Park, and finally Steve Johnson. One relationship we can notice is that the further back the SSA's hire date was (and hence the longer they've been with the company), the greater their total. This is an expected observation, as the amount of the time the Sales Support Agents would have had to garner track sales would be longer too.

Next, we want to analyse the sales data for customers from each different country. For this, we do not look at the billing country that the customer has paid via, and will only look at the registered country of the user themselves.

In particular, we will calculate the following data for each country:

1. Total Number of Customers
2. Total Value of Sales
3. Average Value of Sales per Customer
4. Average Order Value



First, we will create a table of the total number of customers and sales for each country. For simplification, we will group all countries that only have one customer under the 'Other' category.

In [166]:
%%sql


DROP VIEW IF EXISTS total_cust_sales_country;

CREATE VIEW total_cust_sales_country AS

WITH country_value AS (
    SELECT c.country, COUNT(DISTINCT c.customer_id) no_of_customers, SUM(inv.total) total_country, COUNT(inv.invoice_id) total_orders,
            CASE
                WHEN COUNT(DISTINCT c.customer_id) = 1 THEN 'Other'
                ELSE c.country
            END AS cat
    FROM invoice inv
    INNER JOIN customer c ON c.customer_id = inv.customer_id
    GROUP BY c.country
    ORDER BY cat)

    /* Sorting */
    SELECT cat country, SUM(no_of_customers) total_num_customers, ROUND(SUM(total_country), 2) total_value_of_sales, SUM(total_orders) total_orders
    FROM 
        (SELECT *, 
             CASE
             WHEN cat = 'Other' THEN 1
             ELSE 0
             END AS sort
         FROM country_value
        )
    GROUP BY cat
    ORDER BY sort;
    
SELECT *
FROM total_cust_sales_country;


 * sqlite:///chinook.db
Done.
Done.
Done.


[('Brazil', 5, 427.68, 61),
 ('Canada', 8, 535.59, 76),
 ('Czech Republic', 2, 273.24, 30),
 ('France', 5, 389.07, 50),
 ('Germany', 4, 334.62, 41),
 ('India', 2, 183.15, 21),
 ('Portugal', 2, 185.13, 29),
 ('USA', 13, 1040.49, 131),
 ('United Kingdom', 3, 245.52, 28),
 ('Other', 15, 1094.94, 147)]

Next, we check the various statistics that we listed above.

In [167]:
%%sql

SELECT country 'Country', 
        total_num_customers 'Total Number of Customers',
        total_value_of_sales 'Total Value of Sales',
        total_value_of_sales/total_num_customers 'Average Value of Sales per Customer',
        total_value_of_sales/total_orders 'Average Order Value'
        
FROM total_cust_sales_country

 * sqlite:///chinook.db
Done.


[('Brazil', 5, 427.68, 85.536, 7.011147540983607),
 ('Canada', 8, 535.59, 66.94875, 7.047236842105264),
 ('Czech Republic', 2, 273.24, 136.62, 9.108),
 ('France', 5, 389.07, 77.814, 7.7814),
 ('Germany', 4, 334.62, 83.655, 8.161463414634147),
 ('India', 2, 183.15, 91.575, 8.721428571428572),
 ('Portugal', 2, 185.13, 92.565, 6.3837931034482756),
 ('USA', 13, 1040.49, 80.03769230769231, 7.942671755725191),
 ('United Kingdom', 3, 245.52, 81.84, 8.768571428571429),
 ('Other', 15, 1094.94, 72.99600000000001, 7.448571428571429)]

The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:

1. Purchase a whole album
2. Purchase a collection of one or more individual tracks.

The store does not let customers purchase a whole album, and then add individual tracks to that same purchase (unless they do that by choosing each track manually). When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

Management are currently considering changing their purchasing strategy to save money. The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

We have two edge cases to consider:

1. Albums that have only one or two tracks are likely to be purchased by customers as part of a collection of individual tracks.
2. Customers may decide to manually select every track from an album, and then add a few individual tracks from other albums to their purchase.

In the first case, since our analysis is concerned with maximizing revenue we can safely ignore albums consisting of only a few tracks. The company has previously done analysis to confirm that the second case does not happen often, so we can ignore this case also.

We start by writing a query to check if each invoice is an album purchase or not.

In [196]:
%%sql


SELECT i.invoice_id, il.track_id, t.album_id
FROM invoice i
INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id
INNER JOIN track t ON il.track_id = t.track_id
WHERE i.invoice_id = 5;




 * sqlite:///chinook.db
Done.


[(5, 1986, 163),
 (5, 1987, 163),
 (5, 1988, 163),
 (5, 1989, 163),
 (5, 1990, 163),
 (5, 1991, 163),
 (5, 1992, 163),
 (5, 1993, 163),
 (5, 1994, 163),
 (5, 1995, 163),
 (5, 1996, 163),
 (5, 1997, 163),
 (5, 1998, 163),
 (5, 1999, 163),
 (5, 2000, 163),
 (5, 2001, 163),
 (5, 2002, 163)]

In [197]:
%%sql

SELECT a.album_id, t.track_id
FROM album a
INNER JOIN track t ON t.album_id = a.album_id
WHERE a.album_id = 163

 * sqlite:///chinook.db
Done.


[(163, 1986),
 (163, 1987),
 (163, 1988),
 (163, 1989),
 (163, 1990),
 (163, 1991),
 (163, 1992),
 (163, 1993),
 (163, 1994),
 (163, 1995),
 (163, 1996),
 (163, 1997),
 (163, 1998),
 (163, 1999),
 (163, 2000),
 (163, 2001),
 (163, 2002)]

In [3]:
%%sql


SELECT CASE
WHEN ((
SELECT il.track_id
FROM invoice i
INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id
INNER JOIN track t ON il.track_id = t.track_id
WHERE i.invoice_id = 1

EXCEPT

SELECT t.track_id
FROM album a
INNER JOIN track t ON t.album_id = a.album_id
WHERE a.album_id = 92)

AND

(SELECT t.track_id
FROM album a
INNER JOIN track t ON t.album_id = a.album_id
WHERE a.album_id = 92

EXCEPT

SELECT il.track_id
FROM invoice i
INNER JOIN invoice_line il ON i.invoice_id = il.invoice_id
INNER JOIN track t ON il.track_id = t.track_id
WHERE i.invoice_id = 1)) IS NULL THEN 'GAY' ELSE 'NOTGAY' END AS sort


FROM customer

WHERE sort='GAY'

 * sqlite:///chinook.db
Done.


[]

In [29]:
%%sql

DROP VIEW IF EXISTS invoice_album;

CREATE VIEW invoice_album AS
SELECT il.invoice_id, il.track_id invoice_sample_track, t.album_id invoice_sample_album
FROM invoice_line il
INNER JOIN track t ON t.track_id = il.track_id
GROUP BY il.invoice_id;

SELECT *
FROM invoice_album
LIMIT 100;

 * sqlite:///chinook.db
Done.
Done.
Done.


[(1, 1158, 91),
 (2, 3476, 322),
 (3, 2516, 203),
 (4, 3448, 314),
 (5, 1986, 163),
 (6, 1045, 83),
 (7, 3477, 322),
 (8, 2231, 183),
 (9, 206, 21),
 (10, 2663, 215),
 (11, 2597, 211),
 (12, 1463, 118),
 (13, 2553, 206),
 (14, 541, 42),
 (15, 3050, 241),
 (16, 2160, 178),
 (17, 2400, 195),
 (18, 1027, 82),
 (19, 105, 11),
 (20, 905, 72),
 (21, 1461, 118),
 (22, 1000, 80),
 (23, 1, 1),
 (24, 1146, 90),
 (25, 46, 6),
 (26, 2670, 216),
 (27, 3439, 305),
 (28, 36, 5),
 (29, 1025, 82),
 (30, 919, 73),
 (31, 2731, 221),
 (32, 38, 6),
 (33, 3307, 258),
 (34, 610, 49),
 (35, 2377, 194),
 (36, 1582, 127),
 (37, 3458, 321),
 (38, 2960, 234),
 (39, 2308, 189),
 (40, 2382, 194),
 (41, 153, 16),
 (42, 2646, 214),
 (43, 918, 73),
 (44, 3052, 242),
 (45, 1890, 155),
 (46, 2639, 214),
 (47, 3396, 271),
 (48, 2210, 182),
 (49, 2561, 207),
 (50, 1905, 157),
 (51, 2265, 185),
 (52, 2664, 216),
 (53, 1551, 125),
 (54, 1773, 146),
 (55, 1552, 125),
 (56, 2254, 185),
 (57, 2948, 233),
 (58, 1029, 82),
 (59, 449, 37),
 (60, 479, 39),
 (61, 2609, 212),
 (62, 2989, 237),
 (63, 444, 37),
 (64, 38, 6),
 (65, 1139, 89),
 (66, 1435, 116),
 (67, 2401, 195),
 (68, 2575, 209),
 (69, 85, 10),
 (70, 2307, 189),
 (71, 2254, 185),
 (72, 30, 5),
 (73, 2391, 195),
 (74, 1743, 142),
 (75, 22, 4),
 (76, 1020, 82),
 (77, 15, 4),
 (78, 2592, 211),
 (79, 1902, 157),
 (80, 3400, 271),
 (81, 1314, 103),
 (82, 3291, 257),
 (83, 1163, 91),
 (84, 3455, 321),
 (85, 2555, 207),
 (86, 2291, 188),
 (87, 1600, 129),
 (88, 2679, 217),
 (89, 1784, 146),
 (90, 2554, 207),
 (91, 1209, 94),
 (92, 1414, 115),
 (93, 1490, 120),
 (94, 50, 6),
 (95, 3299, 257),
 (96, 1129, 88),
 (97, 1784, 146),
 (98, 1995, 163),
 (99, 2591, 211),
 (100, 996, 79)]

In [20]:
%%sql

SELECT invoice_id, invoice_sample_track, invoice_sample_album
FROM (SELECT *
      FROM invoice inv
      INNER JOIN invoice_album ia ON inv.invoice_id = ia.invoice_id)
LIMIT 100;

 * sqlite:///chinook.db
Done.


[(1, 1158, 91),
 (2, 3476, 322),
 (3, 2516, 203),
 (4, 3448, 314),
 (5, 1986, 163),
 (6, 1045, 83),
 (7, 3477, 322),
 (8, 2231, 183),
 (9, 206, 21),
 (10, 2663, 215),
 (11, 2597, 211),
 (12, 1463, 118),
 (13, 2553, 206),
 (14, 541, 42),
 (15, 3050, 241),
 (16, 2160, 178),
 (17, 2400, 195),
 (18, 1027, 82),
 (19, 105, 11),
 (20, 905, 72),
 (21, 1461, 118),
 (22, 1000, 80),
 (23, 1, 1),
 (24, 1146, 90),
 (25, 46, 6),
 (26, 2670, 216),
 (27, 3439, 305),
 (28, 36, 5),
 (29, 1025, 82),
 (30, 919, 73),
 (31, 2731, 221),
 (32, 38, 6),
 (33, 3307, 258),
 (34, 610, 49),
 (35, 2377, 194),
 (36, 1582, 127),
 (37, 3458, 321),
 (38, 2960, 234),
 (39, 2308, 189),
 (40, 2382, 194),
 (41, 153, 16),
 (42, 2646, 214),
 (43, 918, 73),
 (44, 3052, 242),
 (45, 1890, 155),
 (46, 2639, 214),
 (47, 3396, 271),
 (48, 2210, 182),
 (49, 2561, 207),
 (50, 1905, 157),
 (51, 2265, 185),
 (52, 2664, 216),
 (53, 1551, 125),
 (54, 1773, 146),
 (55, 1552, 125),
 (56, 2254, 185),
 (57, 2948, 233),
 (58, 1029, 82),
 (59, 449, 37),
 (60, 479, 39),
 (61, 2609, 212),
 (62, 2989, 237),
 (63, 444, 37),
 (64, 38, 6),
 (65, 1139, 89),
 (66, 1435, 116),
 (67, 2401, 195),
 (68, 2575, 209),
 (69, 85, 10),
 (70, 2307, 189),
 (71, 2254, 185),
 (72, 30, 5),
 (73, 2391, 195),
 (74, 1743, 142),
 (75, 22, 4),
 (76, 1020, 82),
 (77, 15, 4),
 (78, 2592, 211),
 (79, 1902, 157),
 (80, 3400, 271),
 (81, 1314, 103),
 (82, 3291, 257),
 (83, 1163, 91),
 (84, 3455, 321),
 (85, 2555, 207),
 (86, 2291, 188),
 (87, 1600, 129),
 (88, 2679, 217),
 (89, 1784, 146),
 (90, 2554, 207),
 (91, 1209, 94),
 (92, 1414, 115),
 (93, 1490, 120),
 (94, 50, 6),
 (95, 3299, 257),
 (96, 1129, 88),
 (97, 1784, 146),
 (98, 1995, 163),
 (99, 2591, 211),
 (100, 996, 79)]

In [9]:
%%sql

SELECT invoice_id, invoice_sample_track, invoice_sample_album, track_id
FROM (SELECT *
      FROM invoice inv
      INNER JOIN invoice_album ia ON inv.invoice_id = ia.invoice_id
      INNER JOIN track t ON t.album_id = ia.invoice_sample_album
      )
LIMIT 100;




 * sqlite:///chinook.db
Done.


[(1, 1158, 91, 1158),
 (1, 1158, 91, 1159),
 (1, 1158, 91, 1160),
 (1, 1158, 91, 1161),
 (1, 1158, 91, 1162),
 (1, 1158, 91, 1163),
 (1, 1158, 91, 1164),
 (1, 1158, 91, 1165),
 (1, 1158, 91, 1166),
 (1, 1158, 91, 1167),
 (1, 1158, 91, 1168),
 (1, 1158, 91, 1169),
 (1, 1158, 91, 1170),
 (1, 1158, 91, 1171),
 (1, 1158, 91, 1172),
 (1, 1158, 91, 1173),
 (2, 3476, 322, 3467),
 (2, 3476, 322, 3468),
 (2, 3476, 322, 3469),
 (2, 3476, 322, 3470),
 (2, 3476, 322, 3471),
 (2, 3476, 322, 3472),
 (2, 3476, 322, 3473),
 (2, 3476, 322, 3474),
 (2, 3476, 322, 3475),
 (2, 3476, 322, 3476),
 (2, 3476, 322, 3477),
 (3, 2516, 203, 2506),
 (3, 2516, 203, 2507),
 (3, 2516, 203, 2508),
 (3, 2516, 203, 2509),
 (3, 2516, 203, 2510),
 (3, 2516, 203, 2511),
 (3, 2516, 203, 2512),
 (3, 2516, 203, 2513),
 (3, 2516, 203, 2514),
 (3, 2516, 203, 2515),
 (3, 2516, 203, 2516),
 (3, 2516, 203, 2517),
 (3, 2516, 203, 2518),
 (3, 2516, 203, 2519),
 (3, 2516, 203, 2520),
 (3, 2516, 203, 2521),
 (3, 2516, 203, 2522),
 (4, 3448, 314, 3448),
 (4, 3448, 314, 3492),
 (5, 1986, 163, 1986),
 (5, 1986, 163, 1987),
 (5, 1986, 163, 1988),
 (5, 1986, 163, 1989),
 (5, 1986, 163, 1990),
 (5, 1986, 163, 1991),
 (5, 1986, 163, 1992),
 (5, 1986, 163, 1993),
 (5, 1986, 163, 1994),
 (5, 1986, 163, 1995),
 (5, 1986, 163, 1996),
 (5, 1986, 163, 1997),
 (5, 1986, 163, 1998),
 (5, 1986, 163, 1999),
 (5, 1986, 163, 2000),
 (5, 1986, 163, 2001),
 (5, 1986, 163, 2002),
 (6, 1045, 83, 1033),
 (6, 1045, 83, 1034),
 (6, 1045, 83, 1035),
 (6, 1045, 83, 1036),
 (6, 1045, 83, 1037),
 (6, 1045, 83, 1038),
 (6, 1045, 83, 1039),
 (6, 1045, 83, 1040),
 (6, 1045, 83, 1041),
 (6, 1045, 83, 1042),
 (6, 1045, 83, 1043),
 (6, 1045, 83, 1044),
 (6, 1045, 83, 1045),
 (6, 1045, 83, 1046),
 (6, 1045, 83, 1047),
 (6, 1045, 83, 1048),
 (6, 1045, 83, 1049),
 (6, 1045, 83, 1050),
 (6, 1045, 83, 1051),
 (6, 1045, 83, 1052),
 (6, 1045, 83, 1053),
 (6, 1045, 83, 1054),
 (6, 1045, 83, 1055),
 (6, 1045, 83, 1056),
 (7, 3477, 322, 3467),
 (7, 3477, 322, 3468),
 (7, 3477, 322, 3469),
 (7, 3477, 322, 3470),
 (7, 3477, 322, 3471),
 (7, 3477, 322, 3472),
 (7, 3477, 322, 3473),
 (7, 3477, 322, 3474),
 (7, 3477, 322, 3475),
 (7, 3477, 322, 3476),
 (7, 3477, 322, 3477),
 (8, 2231, 183, 2229),
 (8, 2231, 183, 2230)]

In [50]:
%%sql

DROP VIEW IF EXISTS invoices_of_album_purchases;

CREATE VIEW invoices_of_album_purchases AS
SELECT *
FROM invoice_album ia
WHERE (SELECT il.track_id
       FROM invoice_line il
       WHERE il.invoice_id = ia.invoice_id
      
       EXCEPT
       
       SELECT t.track_id
       FROM track t
       WHERE t.album_id = ia.invoice_sample_album) IS NULL
       
       

       AND
       

      (SELECT t.track_id
       FROM track t
       WHERE t.album_id = ia.invoice_sample_album
       
       EXCEPT
       
       SELECT il.track_id
       FROM invoice_line il
       WHERE il.invoice_id = ia.invoice_id) IS NULL;

SELECT *
FROM invoices_of_album_purchases


 * sqlite:///chinook.db
Done.
Done.
Done.


[(1, 1158, 91),
 (5, 1986, 163),
 (23, 1, 1),
 (24, 1146, 90),
 (31, 2731, 221),
 (32, 38, 6),
 (44, 3052, 242),
 (46, 2639, 214),
 (52, 2664, 216),
 (54, 1773, 146),
 (56, 2254, 185),
 (61, 2609, 212),
 (64, 38, 6),
 (69, 85, 10),
 (71, 2254, 185),
 (73, 2391, 195),
 (76, 1020, 82),
 (79, 1902, 157),
 (84, 3455, 321),
 (90, 2554, 207),
 (92, 1414, 115),
 (99, 2591, 211),
 (112, 1479, 120),
 (114, 1479, 120),
 (123, 2003, 164),
 (138, 2271, 186),
 (141, 2621, 213),
 (142, 23, 5),
 (146, 1466, 119),
 (153, 2406, 196),
 (156, 1882, 155),
 (157, 3336, 260),
 (171, 1466, 119),
 (173, 2285, 188),
 (175, 1755, 145),
 (178, 1102, 87),
 (183, 1033, 83),
 (184, 3409, 278),
 (188, 1455, 118),
 (196, 1325, 105),
 (200, 15, 4),
 (201, 3288, 257),
 (205, 1829, 150),
 (212, 1121, 88),
 (213, 2271, 186),
 (223, 2621, 213),
 (230, 15, 4),
 (237, 2375, 194),
 (247, 2639, 214),
 (253, 1121, 88),
 (257, 2285, 188),
 (262, 77, 9),
 (271, 2271, 186),
 (273, 1479, 120),
 (276, 3300, 258),
 (280, 3081, 244),
 (281, 1902, 157),
 (284, 194, 20),
 (296, 1466, 119),
 (300, 51, 7),
 (305, 2178, 180),
 (308, 2964, 235),
 (317, 3288, 257),
 (320, 3319, 259),
 (330, 1133, 89),
 (334, 1942, 160),
 (335, 2391, 195),
 (341, 1520, 123),
 (344, 1479, 120),
 (362, 695, 55),
 (372, 529, 42),
 (375, 1268, 100),
 (379, 2542, 206),
 (381, 1133, 89),
 (388, 999, 80),
 (389, 3064, 243),
 (392, 1853, 152),
 (398, 2506, 203),
 (399, 1146, 90),
 (404, 1174, 92),
 (413, 2690, 218),
 (414, 1, 1),
 (420, 999, 80),
 (427, 23, 5),
 (444, 2229, 183),
 (447, 2285, 188),
 (449, 1479, 120),
 (457, 1562, 126),
 (465, 149, 16),
 (468, 2149, 178),
 (478, 1479, 120),
 (479, 419, 36),
 (485, 1434, 116),
 (488, 1902, 157),
 (501, 1466, 119),
 (505, 3424, 293),
 (511, 149, 16),
 (512, 807, 65),
 (526, 436, 37),
 (529, 3375, 270),
 (532, 63, 8),
 (533, 2165, 179),
 (537, 1853, 152),
 (540, 3467, 322),
 (556, 2461, 200),
 (578, 2664, 216),
 (581, 3375, 270),
 (582, 891, 72),
 (583, 3467, 322),
 (589, 2271, 186),
 (590, 1362, 109),
 (604, 1755, 145),
 (606, 2003, 164),
 (612, 2204, 182)]

In [55]:
%%sql

DROP VIEW IF EXISTS categorised_purchases;

CREATE VIEW categorised_purchases AS
SELECT *,
       CASE
    
        WHEN (SELECT il.track_id
           FROM invoice_line il
           WHERE il.invoice_id = ia.invoice_id

           EXCEPT

           SELECT t.track_id
           FROM track t
           WHERE t.album_id = ia.invoice_sample_album) IS NULL



           AND


          (SELECT t.track_id
           FROM track t
           WHERE t.album_id = ia.invoice_sample_album

           EXCEPT

           SELECT il.track_id
           FROM invoice_line il
           WHERE il.invoice_id = ia.invoice_id) IS NULL

        THEN 'purchased_album'
        ELSE 'purchased_tracks'
        END AS 'purchase_category'
        
FROM invoice_album ia;

 * sqlite:///chinook.db
Done.
Done.


[]

In [70]:
%%sql

SELECT purchase_category, 
number_of_invoices, 
number_of_invoices/(SELECT CAST(COUNT(*) AS FLOAT) FROM categorised_purchases)*100  pct_of_total_invoices
FROM (
SELECT purchase_category, COUNT(*) number_of_invoices
FROM categorised_purchases
GROUP BY purchase_category
)


 * sqlite:///chinook.db
Done.


[('purchased_album', 114, 18.566775244299674),
 ('purchased_tracks', 500, 81.43322475570032)]

From our table above, we find that a large majority of purchases in the Chinook Store are purchases of individual tracks (81%), compared to the small minority of album purchases (19%). As such, based on this alone, I recommend that we could look at changing our purchasing strategy to focus on purchasing only popular tracks from record companies. However, this could be discussed after further analysis on the amounts that full album purchases bring in vs individual track purchases.